### PR TITLE
fix: fix bug in timSort

### LINKED ISSUE #2687: [BUG]  timsort - std::bad_array_new_length
### Description

Hello all,

For `sorting/tim_sort.cpp`
An exception is thrown: `std::bad_array_new_length`, when you run this piece of code:

```c++
void test(int size) {
    std::cout<<"Try for "<<size<<std::endl;
    int *arr = new int[size];
    for (uint i = 0 ; i < size ; i++) {
        arr[i] = size - i;
    }

    printArray(arr, size);
    timSort(arr, size);
    printf("======\n");
    printArray(arr, size);
    delete arr;
}
```

For example with `test(65);` the exception is thrown.

### PR DESCRIPTION
**Pull request by MaximSmolskiy**
#### Description of Change

Fix #2687 

#### Checklist

- [x] Added description of change
- [x] Added tests and example, test must pass
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Comment by MaximSmolskiy** [comment]
@Panquesito7 @realstealthninja please review

**Comment by MaximSmolskiy** [comment]
@Panquesito7 @realstealthninja please review

File: sorting/tim_sort.cpp [hunk 1]
```diff
@@ -89,8 +91,23 @@ void printArray(int arr[], int n) {
...
+
+    timSort(arr, n);
+    assert(std::is_sorted(arr, arr + n));
+
+    delete[] arr;
+}
+
 // Driver program to test above function
 int main() {
+    test(65);
```
Comment (realstealthninja) [inline sorting/tim_sort.cpp:109]:
```suggestion
    tests();
```

File: sorting/tim_sort.cpp [hunk 2]
```diff
@@ -89,8 +91,23 @@ void printArray(int arr[], int n) {
     std::cout << std::endl;
 }
 
+void test(int n) {
+    int *arr = new int[n];
```
Comment (realstealthninja) [inline sorting/tim_sort.cpp:95]:
```suggestion
    // Case: array of length 65
    int arr[65];
```

File: sorting/tim_sort.cpp [hunk 3]
```diff
@@ -89,8 +91,23 @@ void printArray(int arr[], int n) {
...
+
+    std::iota(arr, arr + n, 0);
+    std::reverse(arr, arr + n);
+    assert(!std::is_sorted(arr, arr + n));
+
+    timSort(arr, n);
+    assert(std::is_sorted(arr, arr + n));
+
+    delete[] arr;
```
Comment (realstealthninja) [inline sorting/tim_sort.cpp:104]:
```suggestion
```

File: sorting/tim_sort.cpp [hunk 4]
```diff
@@ -89,8 +91,23 @@ void printArray(int arr[], int n) {
     std::cout << std::endl;
 }
 
+void test(int n) {
```
Comment (realstealthninja) [inline sorting/tim_sort.cpp:94]:
```suggestion
/**
 * @brief self-test implementation 
 * @returns void
 */
void test() {
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Thank you for the contribution.


The test function should not have a parameter

**Comment by MaximSmolskiy** [comment]
@Panquesito7 @realstealthninja please review

File: sorting/tim_sort.cpp [hunk 5]
```diff
@@ -89,8 +91,27 @@ void printArray(int arr[], int n) {
...
+    std::reverse(arr, arr + N);
+    assert(!std::is_sorted(arr, arr + N));
+
+    timSort(arr, N);
+    assert(std::is_sorted(arr, arr + N));
+}
+
 // Driver program to test above function
 int main() {
+    tests();
```
Comment (realstealthninja) [inline sorting/tim_sort.cpp:113]:
```suggestion
    tests(); // run self test implementations
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Review by realstealthninja — APPROVED** [review]
Thank you for your contributions

**Review by realstealthninja — APPROVED** [review]

**Review by realstealthninja — DISMISSED** [review]

**Comment by realstealthninja** [comment]
unable to merge, will resolve in a day or two! sorry for the delay.

**Comment by realstealthninja** [comment]
since i was the last one to push an update to this pr im unable to merge this. Could you find something to commit to this file?

File: sorting/tim_sort.cpp [hunk 6]
```diff
@@ -89,8 +91,27 @@ void printArray(int arr[], int n) {
...
+    assert(std::is_sorted(arr, arr + N));
+}
+
 // Driver program to test above function
 int main() {
+    tests(); // run self test implementations
+
     int arr[] = {5, 21, 7, 23, 19};
     int n = sizeof(arr) / sizeof(arr[0]);
     printf("Given Array is\n");
```
Comment (realstealthninja) [inline sorting/tim_sort.cpp:117]: could remove the rest of main

File: sorting/tim_sort.cpp [hunk 7]
```diff
@@ -89,8 +91,27 @@ void printArray(int arr[], int n) {
...
+
+    timSort(arr, N);
+    assert(std::is_sorted(arr, arr + N));
+}
+
 // Driver program to test above function
 int main() {
+    tests(); // run self test implementations
+
     int arr[] = {5, 21, 7, 23, 19};
```
Comment (realstealthninja) [inline sorting/tim_sort.cpp:115]: you could remove the rest of main so only what remains is
```cpp
int main() {
   tests(); // run self test implementations
   return 0;
}
```

Comment (MaximSmolskiy) [reply sorting/tim_sort.cpp:115]: @realstealthninja Done, added `const` where it is useful

**Review by realstealthninja — APPROVED** [review]
Looks good to me

**Pull request merged**